File: TigerEx/deposits_withdrawals_and_payments/wallet_system.ts
/**
 * TigerEx Wallet System
 * 
 * Real wallet implementation with deposits, withdrawals, and transfers
 */

export enum WalletType {
  SPOT = 'spot',
  MARGIN = 'margin',
  FUTURES = 'futures',
  EARN = 'earn',
  FEE = 'fee',
  COLLATERAL = 'collateral'
}

export enum TransactionType {
  DEPOSIT = 'deposit',
  WITHDRAWAL = 'withdrawal',
  TRANSFER = 'transfer',
  TRADE = 'trade',
  FEE = 'fee',
  REWARD = 'reward',
  STAKING = 'staking',
  EARN = 'earn',
  REFERRAL = 'referral'
}

export enum TransactionStatus {
  PENDING = 'pending',
  PROCESSING = 'processing',
  COMPLETED = 'completed',
  FAILED = 'failed',
  CANCELLED = 'cancelled'
}

export interface Wallet {
  id: string;
  userId: string;
  type: WalletType;
  asset: string;
  balance: number;
  lockedBalance: number;
  updatedAt: Date;
}

export interface Transaction {
  id: string;
  userId: string;
  walletId: string;
  type: TransactionType;
  asset: string;
  amount: number;
  fee: number;
  status: TransactionStatus;
  txHash?: string;
  address?: string;
  createdAt: Date;
  completedAt?: Date;
}

export interface WithdrawalRequest {
  userId: string;
  asset: string;
  amount: number;
  address: string;
  network: string;
  fee: number;
}

export interface DepositAddress {
  asset: string;
  address: string;
  network: string;
  memo?: string;
  qrCode: string;
}

export class WalletSystem {
  private wallets: Map<string, Wallet> = new Map();
  private transactions: Map<string, Transaction> = new Map();
  private walletIdCounter: number = 0;
  private txIdCounter: number = 0;

  // Initialize user wallets
  async initializeWallets(userId: string, assets: string[]): Promise<Wallet[]> {
    const wallets: Wallet[] = [];
    
    // Create spot wallets for each asset
    for (const asset of assets) {
      const wallet = await this.createWallet(userId, WalletType.SPOT, asset);
      wallets.push(wallet);
    }
    
    // Create special wallets
    wallets.push(await this.createWallet(userId, WalletType.MARGIN, 'USDT'));
    wallets.push(await this.createWallet(userId, WalletType.FUTURES, 'USDT'));
    wallets.push(await this.createWallet(userId, WalletType.EARN, 'USDT'));
    wallets.push(await this.createWallet(userId, WalletType.FEE, 'USDT'));
    
    return wallets;
  }

  // Create wallet
  private async createWallet(userId: string, type: WalletType, asset: string): Promise<Wallet> {
    const wallet: Wallet = {
      id: `WAL-${++this.walletIdCounter}`,
      userId,
      type,
      asset,
      balance: 0,
      lockedBalance: 0,
      updatedAt: new Date()
    };
    
    const key = `${userId}_${type}_${asset}`;
    this.wallets.set(key, wallet);
    return wallet;
  }

  // Get wallet
  getWallet(userId: string, type: WalletType, asset: string): Wallet | undefined {
    const key = `${userId}_${type}_${asset}`;
    return this.wallets.get(key);
  }

  // Get all user wallets
  getUserWallets(userId: string): Wallet[] {
    return Array.from(this.wallets.values())
      .filter(w => w.userId === userId);
  }

  // Get balance
  getBalance(userId: string, type: WalletType, asset: string): number {
    const wallet = this.getWallet(userId, type, asset);
    return wallet ? wallet.balance : 0;
  }

  // Get available balance (total - locked)
  getAvailableBalance(userId: string, type: WalletType, asset: string): number {
    const wallet = this.getWallet(userId, type, asset);
    if (!wallet) return 0;
    return wallet.balance - wallet.lockedBalance;
  }

  // Credit balance
  async credit(userId: string, type: WalletType, asset: string, amount: number, txType: TransactionType, metadata?: Record<string, any>): Promise<Transaction> {
    const wallet = this.getWallet(userId, type, asset);
    if (!wallet) {
      throw new Error('Wallet not found');
    }

    wallet.balance += amount;
    wallet.updatedAt = new Date();

    const transaction = await this.createTransaction(userId, wallet, txType, asset, amount, 0, TransactionStatus.COMPLETED);
    
    return transaction;
  }

  // Debit balance
  async debit(userId: string, type: WalletType, asset: string, amount: number, txType: TransactionType, fee: number = 0): Promise<Transaction> {
    const wallet = this.getWallet(userId, type, asset);
    if (!wallet) {
      throw new Error('Wallet not found');
    }

    const available = wallet.balance - wallet.lockedBalance;
    if (available < amount + fee) {
      throw new Error('Insufficient balance');
    }

    wallet.balance -= (amount + fee);
    wallet.updatedAt = new Date();

    const transaction = await this.createTransaction(userId, wallet, txType, asset, amount, fee, TransactionStatus.COMPLETED);
    
    return transaction;
  }

  // Lock balance (for pending orders)
  async lock(userId: string, type: WalletType, asset: string, amount: number): Promise<void> {
    const wallet = this.getWallet(userId, type, asset);
    if (!wallet) {
      throw new Error('Wallet not found');
    }

    const available = wallet.balance - wallet.lockedBalance;
    if (available < amount) {
      throw new Error('Insufficient available balance');
    }

    wallet.lockedBalance += amount;
    wallet.updatedAt = new Date();
  }

  // Unlock balance
  async unlock(userId: string, type: WalletType, asset: string, amount: number): Promise<void> {
    const wallet = this.getWallet(userId, type, asset);
    if (!wallet) {
      throw new Error('Wallet not found');
    }

    wallet.lockedBalance = Math.max(0, wallet.lockedBalance - amount);
    wallet.updatedAt = new Date();
  }

  // Transfer between wallets
  async transfer(userId: string, fromType: WalletType, toType: WalletType, asset: string, amount: number): Promise<void> {
    // Debit from source
    await this.debit(userId, fromType, asset, amount, TransactionType.TRANSFER);
    
    // Credit to destination
    await this.credit(userId, toType, asset, amount, TransactionType.TRANSFER);
  }

  // Create transaction record
  private async createTransaction(
    userId: string, 
    wallet: Wallet, 
    type: TransactionType, 
    asset: string, 
    amount: number, 
    fee: number,
    status: TransactionStatus
  ): Promise<Transaction> {
    const tx: Transaction = {
      id: `TX-${++this.txIdCounter}`,
      userId,
      walletId: wallet.id,
      type,
      asset,
      amount,
      fee,
      status,
      createdAt: new Date(),
      completedAt: status === TransactionStatus.COMPLETED ? new Date() : undefined
    };
    
    this.transactions.set(tx.id, tx);
    return tx;
  }

  // Get transaction history
  getTransactionHistory(userId: string, limit: number = 50): Transaction[] {
    return Array.from(this.transactions.values())
      .filter(tx => tx.userId === userId)
      .sort((a, b) => b.createdAt.getTime() - a.createdAt.getTime())
      .slice(0, limit);
  }

  // Generate deposit address (simulated)
  async generateDepositAddress(userId: string, asset: string, network: string): Promise<DepositAddress> {
    // In production, integrate with blockchain nodes
    const address = `0x${Date.now().toString(16)}${Math.random().toString(16).slice(2, 42)}`;
    
    return {
      asset,
      address,
      network,
      qrCode: `qr://${address}`,
    };
  }

  // Create withdrawal request
  async createWithdrawal(request: WithdrawalRequest): Promise<Transaction> {
    const wallet = this.getWallet(request.userId, WalletType.SPOT, request.asset);
    if (!wallet) {
      throw new Error('Wallet not found');
    }

    const totalAmount = request.amount + request.fee;
    if (wallet.balance - wallet.lockedBalance < totalAmount) {
      throw new Error('Insufficient balance');
    }

    // Lock amount
    wallet.balance -= totalAmount;
    wallet.lockedBalance += totalAmount;
    wallet.updatedAt = new Date();

    // Create pending transaction
    const tx = await this.createTransaction(
      request.userId,
      wallet,
      TransactionType.WITHDRAWAL,
      request.asset,
      request.amount,
      request.fee,
      TransactionStatus.PENDING
    );
    
    tx.address = request.address;
    tx.txHash = `0x${Date.now().toString(16)}`;
    
    return tx;
  }

  // Confirm withdrawal
  async confirmWithdrawal(txId: string, txHash: string): Promise<void> {
    const tx = this.transactions.get(txId);
    if (!tx || tx.type !== TransactionType.WITHDRAWAL) {
      throw new Error('Transaction not found');
    }

    tx.status = TransactionStatus.COMPLETED;
    tx.txHash = txHash;
    tx.completedAt = new Date();

    // Unlock fee (goes to fee wallet)
    const wallet = this.wallets.get(tx.walletId);
    if (wallet) {
      wallet.lockedBalance -= (tx.amount + tx.fee);
      
      // Credit fee to fee wallet
      const feeWallet = this.getWallet(tx.userId, WalletType.FEE, tx.asset);
      if (feeWallet) {
        feeWallet.balance += tx.fee;
      }
    }
  }

  // Cancel withdrawal
  async cancelWithdrawal(txId: string): Promise<void> {
    const tx = this.transactions.get(txId);
    if (!tx || tx.type !== TransactionType.WITHDRAWAL) {
      throw new Error('Transaction not found');
    }

    if (tx.status !== TransactionStatus.PENDING) {
      throw new Error('Can only cancel pending withdrawals');
    }

    tx.status = TransactionStatus.CANCELLED;
    tx.completedAt = new Date();

    // Refund to wallet
    const wallet = this.wallets.get(tx.walletId);
    if (wallet) {
      wallet.balance += (tx.amount + tx.fee);
      wallet.lockedBalance -= (tx.amount + tx.fee);
    }
  }
}

export default WalletSystem;

File: TigerEx/deposits_withdrawals_and_payments/wallet_system.go
package main

import (
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"time"
)

// Wallet type enumeration
type WalletType string

const (
	WalletTypeSpot     WalletType = "spot"
	WalletTypeMargin   WalletType = "margin"
	WalletTypeFutures  WalletType = "futures"
	WalletTypeEarn    WalletType = "earn"
	WalletTypeFee     WalletType = "fee"
	WalletTypeCollateral WalletType = "collateral"
)

// Transaction type
type TransactionType string

const (
	TypeDeposit    TransactionType = "deposit"
	TypeWithdrawal TransactionType = "withdrawal"
	TypeTransfer  TransactionType = "transfer"
	TypeTrade     TransactionType = "trade"
	TypeFee      TransactionType = "fee"
	TypeReward   TransactionType = "reward"
	TypeStaking  TransactionType = "staking"
	TypeEarn    TransactionType = "earn"
	TypeReferral TransactionType = "referral"
)

// Transaction status
type TransactionStatus string

const (
	StatusPending   TransactionStatus = "pending"
	StatusProcessing TransactionStatus = "processing"
	StatusCompleted TransactionStatus = "completed"
	StatusFailed    TransactionStatus = "failed"
	StatusCancelled TransactionStatus = "cancelled"
)

// Wallet entity
type Wallet struct {
	ID           string     `json:"id"`
	UserID       string     `json:"userId"`
	Type         WalletType `json:"type"`
	Asset        string    `json:"asset"`
	Balance     float64   `json:"balance"`
	LockedBalance float64 `json:"lockedBalance"`
	UpdatedAt   int64     `json:"updatedAt"`
}

// Transaction entity
type Transaction struct {
	ID         string           `json:"id"`
	UserID     string           `json:"userId"`
	WalletID   string           `json:"walletId"`
	Type       TransactionType `json:"type"`
	Asset      string          `json:"asset"`
	Amount     float64         `json:"amount"`
	Fee        float64         `json:"fee"`
	Status     TransactionStatus `json:"status"`
	TxHash     string          `json:"txHash,omitempty"`
	Address    string          `json:"address,omitempty"`
	CreatedAt  int64          `json:"createdAt"`
	CompletedAt *int64        `json:"completedAt,omitempty"`
}

// Withdrawal request
type WithdrawalRequest struct {
	UserID   string  `json:"userId"`
	Asset   string  `json:"asset"`
	Amount  float64 `json:"amount"`
	Address string  `json:"address"`
	Network string  `json:"network"`
	Fee     float64 `json:"fee"`
}

// Deposit address
type DepositAddress struct {
	Asset   string `json:"asset"`
	Address string `json:"address"`
	Network string `json:"network"`
	Memo    string `json:"memo,omitempty"`
	QRCode string `json:"qrCode"`
}

// Wallet system
type WalletSystem struct {
	wallets       map[string]*Wallet
	transactions map[string]*Transaction
}

// NewWalletSystem creates wallet system
func NewWalletSystem() *WalletSystem {
	return &WalletSystem{
		wallets:       make(map[string]*Wallet),
		transactions: make(map[string]*Transaction),
	}
}

// Generate random address
func generateAddress() string {
	b := make([]byte, 20)
	rand.Read(b)
	return "0x" + hex.EncodeToString(b)[:40]
}

// Generate deposit address
func (ws *WalletSystem) GenerateDepositAddress(userID, asset, network string) *DepositAddress {
	addr := generateAddress()
	return &DepositAddress{
		Asset:   asset,
		Address: addr,
		Network: network,
		QRCode:  "qr://" + addr,
	}
}

// Create wallet
func (ws *WalletSystem) CreateWallet(userID, asset string, walletType WalletType) *Wallet {
	wallet := &Wallet{
		ID:             fmt.Sprintf("wallet_%d_%s_%s", time.Now().UnixNano(), userID, asset),
		UserID:         userID,
		Type:           walletType,
		Asset:          asset,
		Balance:       0,
		LockedBalance: 0,
		UpdatedAt:     time.Now().UnixMilli(),
	}
	
	ws.wallets[wallet.ID] = wallet
	return wallet
}

// Get wallet balance
func (ws *WalletSystem) GetBalance(walletID string) float64 {
	wallet, exists := ws.wallets[walletID]
	if !exists {
		return 0
	}
	return wallet.Balance
}

// Get available balance (excluding locked)
func (ws *WalletSystem) GetAvailableBalance(walletID string) float64 {
	wallet, exists := ws.wallets[walletID]
	if !exists {
		return 0
	}
	return wallet.Balance - wallet.LockedBalance
}

// Credit deposit
func (ws *WalletSystem) CreditDeposit(userID, asset, txHash string, amount float64) *Transaction {
	tx := &Transaction{
		ID:        fmt.Sprintf("tx_%d", time.Now().UnixNano()),
		UserID:   userID,
		WalletID: ws.FindWallet(userID, asset, WalletTypeSpot),
		Type:     TypeDeposit,
		Asset:    asset,
		Amount:   amount,
		Fee:      0,
		Status:   StatusCompleted,
		TxHash:   txHash,
		CreatedAt: time.Now().UnixMilli(),
	}
	
	walletID := tx.WalletID
	if wallet, exists := ws.wallets[walletID]; exists {
		wallet.Balance += amount
		wallet.UpdatedAt = time.Now().UnixMilli()
	}
	
	ws.transactions[tx.ID] = tx
	return tx
}

// Process withdrawal
func (ws *WalletSystem) ProcessWithdrawal(req *WithdrawalRequest) *Transaction {
	tx := &Transaction{
		ID:        fmt.Sprintf("tx_%d", time.Now().UnixNano()),
		UserID:    req.UserID,
		WalletID:  ws.FindWallet(req.UserID, req.Asset, WalletTypeSpot),
		Type:      TypeWithdrawal,
		Asset:    req.Asset,
		Amount:   req.Amount,
		Fee:       req.Fee,
		Status:   StatusProcessing,
		Address:  req.Address,
		CreatedAt: time.Now().UnixMilli(),
	}
	
	walletID := tx.WalletID
	if wallet, exists := ws.wallets[walletID]; exists {
		total := req.Amount + req.Fee
		if wallet.Balance >= total {
			wallet.Balance -= total
			wallet.UpdatedAt = time.Now().UnixMilli()
		}
	}
	
	ws.transactions[tx.ID] = tx
	return tx
}

// Transfer between wallets
func (ws *WalletSystem) Transfer(fromUserID, toUserID, asset string, amount float64) (*Transaction, *Transaction) {
	fromWallet := ws.FindWallet(fromUserID, asset, WalletTypeSpot)
	toWallet := ws.FindWallet(toUserID, asset, WalletTypeSpot)
	
	now := time.Now().UnixMilli()
	
	// Debit from sender
	tx1 := &Transaction{
		ID:        fmt.Sprintf("tx_%d", now),
		UserID:    fromUserID,
		WalletID: fromWallet,
		Type:     TypeTransfer,
		Asset:    asset,
		Amount:   amount,
		Status:   StatusCompleted,
		CreatedAt: now,
	}
	
	// Credit receiver
	tx2 := &Transaction{
		ID:        fmt.Sprintf("tx_%d", now+1),
		UserID:    toUserID,
		WalletID: toWallet,
		Type:     TypeTransfer,
		Asset:    asset,
		Amount:   amount,
		Status:   StatusCompleted,
		CreatedAt: now,
	}
	
	// Update balances
	if fw, ok := ws.wallets[fromWallet]; ok {
		fw.Balance -= amount
		fw.UpdatedAt = now
	}
	if tw, ok := ws.wallets[toWallet]; ok {
		tw.Balance += amount
		tw.UpdatedAt = now
	}
	
	ws.transactions[tx1.ID] = tx1
	ws.transactions[tx2.ID] = tx2
	
	return tx1, tx2
}

// Find wallet
func (ws *WalletSystem) FindWallet(userID, asset string, walletType WalletType) string {
	for _, w := range ws.wallets {
		if w.UserID == userID && w.Asset == asset && w.Type == walletType {
			return w.ID
		}
	}
	return ""
}

// Get transaction
func (ws *WalletSystem) GetTransaction(txID string) *Transaction {
	return ws.transactions[txID]
}

// Get user transactions
func (ws *WalletSystem) GetUserTransactions(userID string) []*Transaction {
	var txs []*Transaction
	for _, tx := range ws.transactions {
		if tx.UserID == userID {
			txs = append(txs, tx)
		}
	}
	return txs
}

// Convert to JSON
func (t *Transaction) ToJSON() string {
	b, _ := json.MarshalIndent(t, "", "  ")
	return string(b)
}

func main() {
	ws := NewWalletSystem()
	
	// Create a wallet for user
	wallet := ws.CreateWallet("user123", "BTC", WalletTypeSpot)
	fmt.Printf("Created wallet: %s\n", wallet.ID)
	
	// Simulate a deposit
	tx := ws.CreditDeposit("user123", "BTC", "0xabc123", 1.5)
	fmt.Printf("Deposit TX: %s\n", tx.ToJSON())
	
	// Check balance
	balance := ws.GetBalance(wallet.ID)
	fmt.Printf("Balance: %.8f\n", balance)
	
	// Request withdrawal
	req := &WithdrawalRequest{
		UserID:   "user123",
		Asset:   "BTC",
		Amount:  0.5,
		Address: "0xwithdraw123",
		Network: "Bitcoin",
		Fee:     0.0001,
	}
	withdrawal := ws.ProcessWithdrawal(req)
	fmt.Printf("Withdrawal TX: %s\n", withdrawal.ToJSON())
	
	// Generate deposit address
	depositAddr := ws.GenerateDepositAddress("user123", "BTC", "Bitcoin")
	fmt.Printf("Deposit Address: %+v\n", depositAddr)
}